# Solution Leçon 1 : Introduction à la programmation avec Python

***Il est fortement conseillé d'exécuter toutes les cellules de code, et de bien lire le texte***

## Machines, Programmation, Programmes, et le Programmeur

Les machines, spécialement les appareils digitals (ordinateurs, smartphones, robots) jouent un rôle primordial dans la société moderne. Pour l'essentiel, ces machines sont un assemblage de composants électroniques qui ne peut performer une tache sans l’aide de quelque chose que l'on appelle **programme**. Un ensemble de programmes travaillent en concertation pour nous donner le machine que l'on désigne par le nom "ordinateur".

En gros, ‘un programme demande à un ordinateur de faire quelque chose.’ La création d'un programme exige un individu très spécial que l’on appelle le **programmeur** ou le **codeur** (informellement), qui doit effectuer une activité qui s'appelle la **programmation** ou encore **codage** (informellement). Tout simplement, la programmation est le processus de créer (écrire) une séquence d’instructions qui sera transformée (compilée) en programme qui peut être exécuté par l'ordinateur.

Vous avez, tous et toutes, l'ambition à devenir les grands programmeurs, et, pour y arriver, les programmeurs doivent faire de la programmation! Pour faire de la programmation, nous utilisons des langages de programmation. Les langages de programmation se varient en plusieurs facons. Parmis ces differences il y a une que nous devons comprendre à ce niveau de notre parcours vers notre objectif à devenir les grands programmeurs. Il y a une bifurcation importante qui donne deux types principals de langages de programmation : low-level (bas niveau) et high-level (haut niveau).

### Low-level
Les langages du niveau bas, autrement dit les langages au niveau machine, sont spécifiques à une famille de processeurs. Un langage conçu pour une famille de processeurs ne marchera pas sur un ordinateur doté d'un différent type de processeur. Ces langages sont faits pour contrôler un assemblage de composants électroniques, pas pour être facilement maniable par le cerveau humain. Mais, ils sont nécessaires pour faire marcher un ordinateur et, pour un programmeur qui a appris à penser plus comme une machine qu'un être humain, ils peuvent être utilisés afin d'obtenir un maximum de performance.

### High-level
Les langages de haut niveau sont des langages informatiques qui prennent ses mots-clés d'un langage humain (malheureusement pour vous, ce langage est toujours l'anglais !) et utilisent les conceptions pour le contrôle de la machine qui sont plus compatibles avec le fonctionnement de vos cerveaux. Cela rend la création de programmes plus facile et plus agréable, si vous êtes plus ou moins une personne normale. Vous avez entendu peut-être de plusieurs de ces langages : C, Ruby, Java, et, bien sûr, Python. En fait, ces langages doivent être convertis en code machine, essentiellement en programmes de bas niveau, avant l'exécution, car les ordinateurs ne comprennent que les instructions de bas niveau. Les langages de haut niveau marchent sur presque tous types d'ordinateurs et sur beaucoup de différent appareils digitals comme les smartphones et les systèmes arduino. Bifurcation 2: les langages de haut niveau peuvent être compilés ou interprétés. On explique en bas.
## Machines, Programmation, Programmes, et le Programmeur

Les machines, en particulier les appareils numériques (ordinateurs, smartphones, robots), jouent un rôle primordial dans la société moderne. En essence, ces machines sont un assemblage de composants électroniques qui ne peuvent accomplir une tâche sans l’aide de ce que l’on appelle un **programme**. Un ensemble de programmes travaille en concert pour former la machine que l’on désigne sous le nom "d’ordinateur".

En gros, 'un programme demande à un ordinateur de faire quelque chose'. La création d’un programme nécessite une personne très spéciale que l’on appelle le **programmeur** ou le **codeur** (informellement). Cette personne effectue une activité appelée **programmation** ou **codage** (informellement).
En résumé, la programmation est le processus de création (écriture) d’une séquence d’instructions qui sera transformée (compilée) en un programme exécutable par l’ordinateur.

Vous avez tous et toutes l’ambition de devenir de grands programmeurs. Pour y arriver, les programmeurs doivent pratiquer la programmation ! Pour programmer, nous utilisons des langages de programmation.
Les langages de programmation se distinguent de plusieurs façons. Parmi ces différences, il y en a une que nous devons comprendre à ce stade de notre parcours vers notre objectif : devenir de grands programmeurs.
Il existe une bifurcation importante qui définit deux types principaux de langages de programmation : low-level (bas niveau) et high-level (haut niveau).

### Low-level
Les langages bas niveau, autrement dit les langages au niveau machine, sont spécifiques à une famille de processeurs. Un langage conçu pour une famille de processeurs ne fonctionnera pas sur un ordinateur doté d’un type de processeur différent.
Ces langages sont conçus pour contrôler un assemblage de composants électroniques, et non pour être facilement manipulables par le cerveau humain. Cependant, ils sont indispensables pour faire fonctionner un ordinateur. Pour un programmeur capable de penser plus comme une machine que comme un être humain, ces langages peuvent être utilisés afin d’obtenir un maximum de performance.

### High-level
Les langages de haut niveau sont des langages informatiques dont les mots-clés proviennent d’un langage humain (malheureusement pour vous, ce langage est généralement l’anglais !) et qui utilisent des concepts pour le contrôle de la machine plus compatibles avec le fonctionnement de vos cerveaux. Cela rend la création de programmes plus facile et agréable, surtout si vous êtes une personne “normale”.
Vous avez peut-être entendu parler de plusieurs de ces langages : C, Ruby, Java, et bien sûr Python. En réalité, ces langages doivent être convertis en code machine, essentiellement des programmes bas niveau avant exécution, car les ordinateurs ne comprennent que les instructions de bas niveau.
Les langages de haut niveau fonctionnent sur presque tous les types d’ordinateurs et sur de nombreux appareils numériques, comme les smartphones et les systèmes Arduino.
Bifurcation 2 : les langages de haut niveau peuvent être compilés ou interprétés. Nous l’expliquerons ci-dessous.

#### Langages compilés
Ce sont des langages qui nécessitent un logiciel appelé compilateur. Je sais, c’est un peu étrange d’avoir besoin d’un logiciel pour créer un autre logiciel, mais c’est ainsi.
Le compilateur transforme nos instructions joliment rédigées en langage presque humain en un programme que l’ordinateur peut exécuter.
Des exemples de langages compilés sont : C, C++, Java* (Java est un cas un peu particulier, mais nous attendrons le cours de Java pour entrer dans les détails).
Voici le schéma illustrant le fonctionnement d’un compilateur :

>>> fichier de code → compilateur → fichier d’execution → resulat du programme

#### Langages interpretés
Ce sont des langages qui utilisent un interpréteur, encore un logiciel, pour exécuter le programme. Un interpréteur et un compilateur accomplissent à peu près la même tâche : transformer nos instructions en langage machine.
La différence est que l’interpréteur lit les instructions une par une, exécutant immédiatement leur équivalent en langage machine, alors que le compilateur lit toutes les instructions du fichier programme, puis produit un programme en langage machine globalement optimisé.
Les langages compilés sont généralement plus performants que les langages interprétés, mais il est plus facile de développer et de tester des programmes avec un interpréteur.
Python est un exemple de langage interprété. Dans notre cas, nous utiliserons Jupyter Notebook comme interpréteur (en réalité, l’interpréteur est IPython, intégré à Jupyter Notebook).
Voici le schéma illustrant le fonctionnement d’un interpréteur :

>>> fichier de code → interpreteur → resulat du program

#### Langages compilés
C'est des langages qui ont besoin d'un logiciel qui s'appelle un compilateur. Je sais, c'est bizarre que l'on ait besoin d'un logiciel pour créer le logiciel, mais c'est comme ça. Ce fameux compilateur transforme nos instructions joliment écrites en langage presque humain en un programme que l'ordinateur peut exécuter. Des exemples de langages compilés sont : C, C ++, Java * (le Java est un peu un cas spécial, mais il faut attendre pour le cours de Java pour entrer dans les détails). Voici le schéma pour l'opération d'un compilateur :

>>> fichier de code → compilateur → fichier d’execution → resulat du programme

#### Langages interpretés
C'est des langages qui utilisent un interpréteur, encore un logiciel, pour exécuter le programme. Un interpréteur et un compilateur font un peu près la même chose, transforment nos instructions en langage machine. La différence est que l'interpréteur lit les instructions une par une, exécutant son équivalent en langage machine sur le champ alors que le compilateur lit toutes les instructions écrites dans le fichier programme, puis donne le programme en langage machine globalement optimisé comme résultat. Les langages compilés sont plus performants que les langages interprétés, mais il est plus facile de développer et de tester les programmes quand on utilise un interpréteur. Python est un exemple de langage interprété. Dans notre cas, nous utiliserons Jupyter Notebook comme interpréteur (en réalité, l'interpréteur est IPython, intégré en Jupyter Notebook). Voici le schéma pour l'opération d'un interpréteur :

>>> fichier de code → interpreteur → resulat du program

## Python et le cours

### Qu'est ce que Python ?

- ##### Une langue de programmation à l'usage général
- ##### Un langage interpreté -> cela veut dire qu'on saisit des commandes et on reçoit une réponse directement
- ##### Elle est relativement simple et moins difficile pour les débutants que beaucoup d'autres langages

###  Que pouvons-nous faire avec Python ?

- ##### Programmation scientifique
- ##### Robotique
- ##### Développement de jeux video
- ##### Analyse de données
- ##### Développement d'applications pour le Web
- ##### Administration système
- ##### Vision par ordinateur &amp; Réalite Augmentée
- ##### Intelligence Artificielle / Apprentissage Machine
- ##### Tout ce que l'on veut. La seule limite est ton imagination !

### Quels sont les objectifs de ce cours en Python ?

- ##### (primaire) Apprendre à résoudre les problèmes en utilisant Python
- ##### Apprendre à programmer
- ##### Ecrire des programmes sympas
- ##### Rejoindre le mouvement pour numériser le Mali (et devenir un citoyen utile!)
- ##### S'amuser !!!
## Python et le cours

### Qu'est ce que Python ?

- Un langage de programmation à usage général.
- Un langage interprété → cela signifie que l’on saisit des commandes et que l’on reçoit une réponse immédiatement.
- Il est relativement simple et plus facile pour les débutants que beaucoup d’autres langages.

###  Que pouvons-nous faire avec Python ?

- Programmation scientifique
- Robotique
- Développement de jeux vidéo
- Analyse de données
- Développement d’applications web
- Administration système
- Vision par ordinateur, réalité virtuelle &amp; réalité augmentée 
- Intelligence artificielle / Apprentissage automatique
- Tout ce que tu veux : la seule limite est ton imagination !

### Quels sont les objectifs de ce cours en Python ?

- (Primaire) Apprendre à résoudre des problèmes en utilisant Python
- Apprendre à programmer
- Écrire des programmes sympas
- Rejoindre le mouvement pour numériser le Mali (et devenir un citoyen utile !)
- S’amuser !

## Le programme de 'Hello, World'

In [1]:
# Exemple1
print("Hello, World")

Hello, World


Ceci est notre premier programme de Python. Facile, non ?  

Ici, tu as demandé à Jupyter Notebook d’écrire les mots **Hello, World** sur l’écran.

Essayons d’analyser cette programme:

- **print()**: est un fonction de base intégré dans le langage Python. Cela veut dire ??? tu dis. Pas de souci, tu vas apprendre tout sur les fonctions, mais pas toute de suite. Pour l'instant, nous pouvons dire que c'est une commande qui dit a l’ordinateur d’écrire ce qui est entre les parentheses. Les parentheses fait parti de la commande,  il ne faut pas les omettre!

- **"Hello, World"**: C’est le texte que nous voulons écrire sur l’écran. Ou, plûtot, c'est ce que nous voulons écrire sauf les guillemets. En Python, les guillements anglais créent une chaine de caracteres, comme les fonctions, quelque chose que tu vas bien connaitre et peut-être aimer, mais, pour l'instant, il faut juste savoir que, comme les parentheses, ils sont obligatoire.

In [2]:
# Exemple2
print("Salut, le monde")

Salut, le monde


Tu peux changer le texte "Hello, World", toutes les lettres et les chiffres sont possible, ainsi que l'apostrophe, le point et les autres marques de ponctuation. Exécuter :

In [3]:
# Exemple3
print("J'ai 15 ans.")

J'ai 15 ans.


Si tu peux taper un symbole avec ton clavier, tu peux le mettre entre guillemets et puis l'imprimer à l'écran avec print. Exécuter : 

In [4]:
# Exemple4
print("&é'(-è_çà)=~#{[|`\^@]}$€µ*ù%ɛƐɔƆŋŊɲƝ§!:/;.,?")

&é'(-è_çà)=~#{[|`\^@]}$€µ*ù%ɛƐɔƆŋŊɲƝ§!:/;.,?


Sont les guillemets toujours requiss ? Voyons : 

In [5]:
# Exemple5
print(120)

120


En Python, les nombres sont accepté avec print sans d'être entre les guillemets. Mais, qu'est-ce qu'il y a une différence entre un nombre entre guillemets et un nombre sans guillemets ? Essayons : 

In [6]:
# Example6
print("42")
print(42) 

42
42


Tu pourrais penser que les résultats sont les mêmes ... mais tu aurais tort. Peut-être, tu as appris la différence entre un chiffre et un nombre. Effectivement, c'est ça. En Python, ce qui est entre guillemets sera une chaîne de caractères qui pourrait être les lettres ou les chiffres. Sans guillemets, une valeur numérique est un nombre, cela veut dire, une quantité avec laquelle ton programme peux calculer. Tu vas comprendre cela mieux en quelques instants.

## Le programme 'Simple Calculatrice'

C'est facile de faire les calculs en Python avec rien plus que notre ami, la fonction **print** et les opérateurs classique de la mathématique. Voyons :

### Addition

In [7]:
# Exemple7
print(25+25)

50


C'est comme ça que tu fais l'addition avec Python. Simple, non ? Et pour avoir le résultat, il ne faut qu'exécuter la cellule.

### Soustraction

In [8]:
# Exemple8
print(25-10)

15


Comme addition, la soustraction ne te posera pas un casse-tête. Avec juste un petite signe de soustraction entre deux nombres, et voila! tu auras le résultat que tu cherches.

In [9]:
# Exemple9
print(10-25)

-15


Un résultat négatif ! Pas de souci, en Python, tu peux aussi faire des opérations qui donnent les résultats négatifs.

### Multiplication

In [10]:
# Exemple10
print(512*2)

1024


On fait la multiplication en utilisant l'asterisque (**\***).

In [11]:
# Exemple11
print(2*4-4+2)

6


Tu peux enchaîner les opérations en Python, comme c'est fait dans l'Exemple11, où on trouve la multiplication, la soustraction, et l'addition.

### Division

Pour se remuer les méninges, calcule le résultat suivant avant que tu demandes Python de le faire pour toi.

In [12]:
# Exemple12
print(47/14)

3.357142857142857


On fait division en utilisant la barre oblique (slash) **/**.

C'est quoi ce résultat? Un nombre décimale? Oui, il y a les nombres décimaux en Python.

Calcule le résultat de l'opération suivante avant que tu l'exécute.

In [13]:
# Exemple13
print(5*4+5/5)

21.0


As-tu trouvé le bon résultat ? Si non, le problème n'est pas que tu es nul en maths, ou j'espère que ce n'est pas le cas, mais parce que toi et Python ne sont pas d'accord sur le bon ordre des opérations. Tu peux utiliser les parenthèses pour fixer le groupement d'opérations, comme dans l'exemple en bas. Calcule encore le résultat. Je parie que cette fois-ci, tu as trouvé le bon résultat. Ou non ? Si non, peut-être, c'est l'heure de réviser tes maths.

In [14]:
# Exemple14
print(((5*4)+5)/5)

5.0


### Elevation à la puissance (l'opérateur exposant)

In [15]:
# Exemple15
print(7**2)

49


Cela peut sembler bizarre, mais l'élévation à la puissance est fait avec **\*\***. C'est parce qu'il n'est pas évident de taper **7²** sur le clavier, **7\*\*2** est donc plus agréable.

C'est assez pour le moment. Comme nous disons en anglais, we've only scratched the surface (à peine amorcé le travail), mais Rome wasn't built in a day (Rome ne s'est pas construite en un jour). Cela sera bien de prendre du temps maintenant de jouer avec les opérations mathématiques, print, et les chaînes de caractères.
C’est suffisant pour le moment. Comme on dit en anglais, we’ve only scratched the surface (à peine amorcé le travail), mais Rome wasn’t built in a day (Rome ne s’est pas construite en un jour).
Il sera bien de prendre un peu de temps maintenant pour jouer avec les opérations mathématiques, print, et les chaînes de caractères.

## Programmation et les Bugs (les erreurs de programmation, les bouges)

In every life, a little rain must fall (Dans toute vie, il faut que tombe un peu de pluie). Dans la programmation, les bugs sont la pluie, mais, en vérité, c'est pas **un peu** mais une inondation à la malienne ! Tu ne feras jamais la programmation pas sans se heurter à un déluge de bugs. Alors, c'est quoi, ces fameux bugs dont on parle ? Les bugs sont les erreurs que tu rencontreras quand tu es train de créer tes programmes et quand tu les teste ou, pire encore, quand tes clients utilisent le logiciel que tu les as vendu ! Mieux trouver les bugs tôt plutôt tard ! Tu ne seras jamais un programmeur sans ayant appris à comprendre les bugs et les tuer comme les cafards !

En programmation, il y a trois types d'erreurs que tu vas rencontrer :

### Erreur de syntaxe (Syntax Error)

Les langages de programmation ont une structure et des règles qui sont bien définis. Et Python fait pas exception, quand tu ne respectes pas les règles de Python, Python ne te traitera pas gentiment !

In [16]:
# Exemple16
print(2+4

SyntaxError: unexpected EOF while parsing (<ipython-input-16-e0567bf4f207>, line 2)

Ceci est notre premier exemple d'erreur lié à la syntaxe. Les parenthèses sont comme maman et les portes - si tu ouvres la porte, il faut la fermer après. En Python, quand tu ouvres un bloc de parenthèse avec (, il faut plus tard le fermer avec ).

Les erreurs de syntaxe sont signalées immédiatement quand l'interpréteur essaye d'exécuter le programme.

Une compréhension de base de l'anglais est plus qu'utile pour comprendre ces erreurs. Si l'anglais n'est pas encore ton point fort, il faut, quand même, essayer de lire le message. Tu vas voir les mêmes messages encore et encore, et avec le temps, tu vas arriver à comprendre ce que le message veut dire. Ou non, nous n'avons pas dit au début que l'anglais est incontournable pour le travail technique ? 

Le format du message d'erreur de syntaxe est le suivant :

- Un ligne avec le nom du fichier, et le nombre de la ligne où l'erreur s'est produit.

- Un copie de la ligne où l'erreur s'est produit, avec **^** qui indique le point exact où l'erreur a été détecté (l'erreur peut être là ou avant cette position)

- Un ligne qui donne le type d'erreur, suit par un deux points(**:**) et une description de l'erreur.

### Erreur d'exécution (Runtime Error)

In [17]:
# Exemple17
print(25/0)

ZeroDivisionError: division by zero

Avec une erreur d'exécution, tu n'as pas enfreint les règles de syntaxe de Python, mais, quand même, tu as commis une bavure que Python, une fois en cours de l'exécution de ton programme, ne peut pas supporter. Le programme arrêtera quand il arrive à la ligne fautive. Entre amis, peut-être, on peut laisser passer une bavure, mais Python ne te pardonnera pas ! Ces erreurs sont aussi appelés **exception**. Il y a aussi les exceptions que le programmeur peut définir afin d'arrêter le programme quand quelque chose bizarre arrive, mais ce type d'exception n'est pas nos oignons pour l'instant. 

Dans notre exemple, bien que la syntaxe soit correcte, on a signalé une erreur. C'est parce que on a divisé un numéro par zéro et ni toi, ni Python, ni le meilleur mathématicien du monde sait ce que cela donne. Python te dit simplement que ce que tu l'as demandé de faire est idiot.

Le format de message de cette erreur est le même qu'une erreur de syntaxe.

Une liste complète des exceptions de Python se trouve ici :

https://docs.python.org/fr/3/library/exceptions.html#bltin-exceptions

Nous allons traiter les exceptions plus tard dans le cours jusque tu les en auras marre. 

### Erreur de semantique

In [18]:
# Exemple18
print(12+4*2/3)

14.666666666666666


C'est les erreurs qui sont les plus difficiles à répérer parce qu'ils ne sont pas en réalité les erreurs. Tu as bien demandé Python de faire quelque chose et Python l'a fait fidèlement, mais tu n'as pas trouvé le résultat que tu as attendu. Pourquoi ? Parce que tu n'as pas donné les instructions qu'il faut à Python. 

Par exemple, dans Exemple18, disons que nous voulons 10.66 comme résultat, mais avec notre code nous avons 14.67. C'est correcte, mais pas ce que nous avons attendu. Peux-tu corriger cette erreur de sémantique ?

Ici, contrairement à ce que nous avons dit, l'erreur est facile à repérer. Mais il faut imaginer que l'erreur sémantique avec le mauvais calcul est enfouies dans des centaines de lignes de code et on remarque le résultat erroné à la fin, sans aucune idée de l'origine du problème. On sera obligé de vérifier ligne par ligne la bonne exécution de programme. Ca, c'est l'enfer de programmeurs !
Ce sont les erreurs les plus difficiles à repérer, car elles ne sont pas vraiment des erreurs.
Tu as bien demandé à Python de faire quelque chose, et Python l’a fait fidèlement, mais le résultat n’est pas celui que tu attendais.
Pourquoi ? Parce que tu n’as pas donné les instructions correctes à Python.

Par exemple, dans Exemple 18, imaginons que nous voulons obtenir 10,66 comme résultat, mais notre code renvoie 14,67. C’est correct du point de vue syntaxique et exécutable, mais ce n’est pas ce que nous attendions. Peux-tu corriger cette erreur de sémantique ?

Ici, contrairement aux erreurs précédentes, l’erreur est facile à repérer.
Mais imagine que cette erreur sémantique, liée à un mauvais calcul, soit enfouie dans des centaines de lignes de code : on ne remarque le résultat erroné qu’à la fin, sans aucune idée de l’origine du problème.
Il faudra alors vérifier ligne par ligne pour s’assurer que le programme fonctionne correctement.
Voilà l’enfer des programmeurs !

In [1]:
# Correction d'Exemple18
print((12+4)*2/3)

10.666666666666666


Ouuff! Tu es bien arrivé à la fin du leçon, avant de faire les exercices, faisons un petit récapitulatif.

## Récapitulatif

- On fait un programme pour travailler sur les machines digitales, les programmes sont créés lors du processus de la programmation, et un programmeur est celui qui fait la programmation avec un langage de programmation.

- On a plusieurs types de langages de programmation. Ils sont classifiés en deux types: low-level, high-level

- Les langages high-level sont encore classifiés en deux types : les langages compilés, et les langages interprétés

- Python est un langage interprété.

- On écrit quelque chose sur l'écran avec la commande **print()** de Python

- Les chaînes de caractères sont créées par le placement d'une séquence de caractères (lettres, chiffres, marques de ponctuation, etc.) entre guillemets anglais ("")

- Les opérations mathématiques sont faits avec +, -, *, /, **, ce qui est l'addition, la soustraction, la multiplication, la division, et l'élevation à la puissance respectivement

- On a trois types d'erreurs en programmation :
    - Erreur de syntaxe
    - Erreur d'exécution (exception)
    - Erreur de semantique
    Un programme est créé pour faire fonctionner les machines digitales. Les programmes sont créés lors du processus de programmation, et un programmeur est celui qui réalise cette programmation à l’aide d’un langage de programmation.

- Il existe plusieurs types de langages de programmation, classés en deux catégories : low-level (bas niveau) et high-level (haut niveau).

- Les langages high-level se divisent encore en deux types : langages compilés et langages interprétés.

- Python est un langage interprété.

- Pour afficher quelque chose à l’écran, on utilise la commande **print()** de Python

- Les chaînes de caractères sont créées en plaçant une séquence de caractères (lettres, chiffres, ponctuation, etc.) entre guillemets anglais ("").

- Les opérations mathématiques se font avec :

+ : addition
- : soustraction
* : multiplication
/ : division
** : élévation à la puissance

- On distingue trois types d’erreurs en programmation :
    - Erreur syntaxiaue
    - Erreur d'exécution (exception)
    - Erreur sémantique

## Problèmes

**Il faut exécuter la cellule ci-dessous avant de continuer. Si tu ne suis pas cette instruction, les exercices qui suivent ne marcheront pas ! **

Si tu fermes le notebook avant de terminer les exercices, ou si tu voudrais changer tes réponses, il faut exécuter encore la cellule.

In [19]:
verificateur = lambda x, y: 'Bien' if x == y else 'Oops, manqué! Reessayer.'

- Dans la cellule de code ci-dessous, écris 'Je suis __ et j'ai _ ans.' sur l'écran. Remplace les espaces vides avec ton nom et ton âge.

In [20]:
# Probleme1
print("Je suis Jean Doe et j'ai 15 ans.")

Je suis Jean Doe et j'ai 15 ans.


- Dans le programme suivant, remplace l'espace vide avec ton classe.

In [21]:
# Probleme2
print("Je fais la 10e année.")

Je fais la 10e année.


- Dans le programme suivant, entre le code Python qui manque pour écrire le message sur l'écron

In [22]:
# Probleme3
print("J'apprend la programmation en Python avec RobotsMali")

J'apprend la programmation en Python avec RobotsMali


- Ecris une phrase qui contient des lettres, au moins un chiffre et une ponctuation

In [23]:
# Probleme4
print("Python3 est fun!")

Python3 est fun!


- Remplace l'espace vide avec l'operation 8 multiplié par 2

In [24]:
# Probleme5
verificateur(8*2, 16)

'Bien'

- Remplace l'espace vide avec l'opération 10 divisé par 5 avant la virgule et le résultat que cela donne apres la virgule. 

In [25]:
# Probleme6
verificateur(10 / 5, 2)

'Bien'

- Qu'est ce qui manque dans les opérations ci-dessous ?

In [26]:
# Probleme7a
verificateur(150 - 75, 75)

'Bien'

In [27]:
# Probleme7b
verificateur(2 ** 5, 32)

'Bien'

- Quel nombre soustrait de 956 donne 167 ?

In [28]:
# Probleme8
verificateur(956 - 789, 167)

'Bien'

- Le carré de quel nombres manque dans les opérations suivantes ?

In [29]:
# Probleme9a
verificateur(13 ** 2, 169)

'Bien'

In [30]:
# Probleme9b
verificateur(9 ** 2, 81)

'Bien'

In [31]:
# Probleme9c
verificateur(20 ** 2, 400)

'Bien'

In [32]:
# Probleme9d
verificateur(15 ** 2, 225)

'Bien'

In [33]:
#Probleme9e
verificateur(64 ** 2, 4096)

'Bien'

### Tuer les Bugs

Régler les erreurs dans les programmes au-dessous.

In [34]:
# Bug1
print("Voici un bug")

Voici un bug


In [35]:
# Bug2
print("Je suis un bug")

Je suis un bug


In [36]:
# Bug3
print("Un bug se cache ici")

Un bug se cache ici


In [37]:
# Bug4
print('Me voilà, un bug')

Me voilà, un bug


In [38]:
# Bug5
print(2+30)

32


In [39]:
# Bug6
print(3/1)

3.0


Fait ce qu'il faut faire afin de réussir la vérification. (Les codes avant et après la virgule doivent avoir les valeurs égales.)

In [40]:
# Bug7
verificateur((50 + 2) * 5, 260)

'Bien'

In [41]:
# Bug8
verificateur("Hello, Mali", "Hello, Mali")

'Bien'

In [42]:
# Bug9
verificateur("Je programme", "Je programme")

'Bien'

In [43]:
# Bug10
verificateur(18 / (6 * 3), 1)

'Bien'

In [44]:
# Bug11
verificateur(6 ** 2, 36)

'Bien'

**Félications! You are off and running (tu as le vent dans les voiles), c'est terminé ton introduction à Python. En avant!**